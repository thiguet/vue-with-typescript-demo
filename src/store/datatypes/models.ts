export enum Measures {
    UN = 'UN',
    L = 'L',
    KG = 'KG',
    G = 'G',
}

export interface Product {
    name: string;
    measure: keyof typeof Measures;
    qtd: number;
    minQtd: number;
    img?: string;
}

export interface User {
    name: string;
    email: string;
}

export interface LoginSubmit {
    name: string;
    pass: string;
}

export enum VuexAppModules {
    products = 'products',
    login = 'login',
    alert = 'alert',
}
