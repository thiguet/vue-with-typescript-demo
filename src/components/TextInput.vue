<template>
    <input
        :id="id"
        :type="type"
        :name="id"
        class="input"
        :value="value"
        @input="onInput"
    />
</template>

<script lang="ts">
import Vue from 'vue';

import Component from 'vue-class-component';
import { Prop } from 'vue-property-decorator';

@Component({
    name: 'TextInput',
})
export default class TextInput extends Vue {
    @Prop({ required: true }) readonly id!: string;

    @Prop({ default: 'text' }) readonly type!: string;

    @Prop({ required: true }) readonly value!: string;

    @Prop({ required: true }) readonly setValue!: Function;

    private onInput(e: Event) {
        this.setValue((e.target as HTMLInputElement).value);
    }
}
</script>

<style scoped>
input {
    border: none;
    border-bottom: 1px solid var(--border-color);
    width: 100%;
    font-family: inherit;
    font-size: var(--font-input-size);
}
</style>
