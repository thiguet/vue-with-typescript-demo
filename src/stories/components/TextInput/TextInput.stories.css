@import url('../../app.css');

.input {
    border-bottom: 1px solid var(--border-color) !important;
}
