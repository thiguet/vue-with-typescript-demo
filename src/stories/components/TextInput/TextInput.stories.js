import TextInput from '@/components/TextInput.vue';
import './TextInput.stories.css';

export default {
    title: 'Example/Components/TextInput',
    component: TextInput,
};

const Template = (_args, { argTypes }) => ({
    props: Object.keys(argTypes),
    components: { TextInput },
    template: '<text-input v-bind="$props"/>',
});

let value = 'someVal';

const setValue = (val) => {
    value = val;
};

export const Default = Template.bind({});
Default.args = {
    id: 'some-id',
    value,
    setValue,
};
