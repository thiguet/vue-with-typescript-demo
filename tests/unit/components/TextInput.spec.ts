import { shallowMount } from '@vue/test-utils';
import TextInput from '@/components/TextInput.vue';
import faker from 'faker';
import { TextInputProps } from './models';

describe('TextInput.vue', () => {
    let props: TextInputProps;

    const build = () => {
        const wrapper = shallowMount(TextInput, {
            propsData: { ...props },
        });
        const input = () => wrapper.find('.input');

        return {
            wrapper,
            input,
            inputEl: (): HTMLInputElement => input().element as HTMLInputElement,
        };
    };

    beforeEach(() => {
        props = {
            id: faker.random.alphaNumeric(),
            type: 'text',
            value: faker.random.word(),
            setValue: jest.fn(),
        };
    });

    it('renders component', () => {
        props.id = 'someFixedValidId';
        props.value = 'someFixedValidValue';

        const { wrapper } = build();
        expect(wrapper).toMatchSnapshot();
    });

    it('renders main child components', () => {
        const { input } = build();

        expect(input().exists()).toBe(true);
    });

    it('passes props to the component', () => {
        const { wrapper } = build();

        const { id, type, value, setValue } = wrapper.props();

        expect(id).toBe(props.id);
        expect(type).toBe(props.type);
        expect(value).toBe(props.value);
        expect(setValue).toBe(props.setValue);
    });

    it("sets input's type propety to element by props", () => {
        const { wrapper, input } = build();

        const { type } = wrapper.props();

        expect(input().attributes().type).toBe(type);
    });

    it('sets type to text using default value', () => {
        delete props.type;

        const { wrapper } = build();

        expect(wrapper.props().type).toBe('text');
    });

    it("sets input's input propety to setValue fn", async () => {
        const { wrapper, inputEl, input } = build();

        const { setValue } = wrapper.props();

        const value = faker.random.word();

        inputEl().value = value;

        await input().trigger('input');

        expect(setValue).toHaveBeenCalledWith(value);
    });
});
