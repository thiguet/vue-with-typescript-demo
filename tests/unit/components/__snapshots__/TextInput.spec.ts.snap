// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput.vue renders component 1`] = `<input id="someFixedValidId" type="text" name="someFixedValidId" class="input">`;
